File: artifact.go
package main

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"unicode"
)

var (
	doiRE      = regexp.MustCompile(`10\.\d{4,9}/[-._;()/:A-Za-z0-9]+`)
	filenameRE = regexp.MustCompile(`(?i)\b(?:[\w.-]+/)*[\w.-]+\.(py|r|js|ts|go|rs|ipynb|sh|rb|jl|txt)\b`)
	fenceRE    = regexp.MustCompile("(?s)```(?:[A-Za-z0-9_+-]+)?\\s*\\n(.*?)```")
)

func looksLikeArtifactRequest(text string) bool {
	q := strings.ToLower(text)
	if strings.TrimSpace(q) == "" {
		return false
	}
	verbs := []string{"write", "create", "generate", "make", "draft", "scaffold", "emit", "produce"}
	nouns := []string{"script", "python", "chempy", "notebook", "module", "program", "source file", ".py", "code"}
	hasVerb := false
	for _, verb := range verbs {
		if strings.Contains(q, verb) {
			hasVerb = true
			break
		}
	}
	hasNoun := false
	for _, noun := range nouns {
		if strings.Contains(q, noun) {
			hasNoun = true
			break
		}
	}
	if hasVerb && hasNoun {
		return true
	}
	if strings.Contains(q, "write the script") || strings.Contains(q, "script now") {
		return true
	}
	if filenameRE.FindString(text) != "" && hasVerb {
		return true
	}
	return false
}

func researchQueryFrom(userText, transcript string) string {
	cleaned := stripArtifactBoilerplate(userText)
	if wordCount(cleaned) < 2 {
		cleaned = stripArtifactBoilerplate(transcript)
	}
	if wordCount(cleaned) < 2 {
		cleaned = keepMedicalTerms(userText + " " + transcript)
	}
	if doi := doiRE.FindString(transcript + " " + userText); doi != "" {
		cleaned = strings.TrimSpace(cleaned + " " + strings.TrimRight(doi, ".,;"))
	}
	cleaned = strings.Join(strings.Fields(cleaned), " ")
	if cleaned == "" {
		return strings.TrimSpace(userText)
	}
	return cleaned
}

func stripArtifactBoilerplate(text string) string {
	stop := map[string]bool{
		"write": true, "writing": true, "create": true, "creating": true, "generate": true,
		"generating": true, "make": true, "making": true, "draft": true, "scaffold": true,
		"python": true, "chempy": true, "script": true, "scripts": true, "file": true,
		"files": true, "code": true, "program": true, "now": true, "please": true,
		"based": true, "research": true, "the": true, "a": true, "an": true, "to": true,
		"for": true, "begin": true, "beginning": true, "start": true, "starting": true,
		"do": true, "that": true, "this": true, "with": true, "using": true, "use": true,
		"want": true, "wants": true, "user": true, "preprint": true, "passage": true,
		"passages": true, "therefore": true, "answer": true, "from": true, "found": true,
		"in": true, "of": true, "and": true, "or": true, "on": true, "it": true,
		"is": true, "be": true, "can": true, "should": true, "able": true, "just": true,
		"so": true, "those": true, "these": true, "i": true, "will": true, "get": true,
		"numpy": true, "pandas": true, "scipy": true, "jupyter": true, "notebook": true,
	}
	var kept []string
	for _, tok := range strings.Fields(strings.ToLower(text)) {
		tok = strings.Trim(tok, ".,;:!?\"'`")
		if tok == "" || stop[tok] {
			continue
		}
		if filenameRE.MatchString(tok) {
			continue
		}
		kept = append(kept, tok)
	}
	return strings.Join(kept, " ")
}

func keepMedicalTerms(text string) string {
	hints := []string{
		"cancer", "vaccine", "mrna", "rna", "immun", "oncolog", "tumor", "tumour",
		"antigen", "peptide", "trial", "therap", "patient", "clinic", "epidem",
	}
	q := strings.ToLower(text)
	var kept []string
	seen := map[string]bool{}
	for _, tok := range strings.Fields(q) {
		tok = strings.Trim(tok, ".,;:!?\"'`")
		for _, hint := range hints {
			if strings.Contains(tok, hint) && !seen[tok] {
				kept = append(kept, tok)
				seen[tok] = true
				break
			}
		}
	}
	return strings.Join(kept, " ")
}

func wordCount(text string) int {
	return len(strings.Fields(text))
}

func inferArtifactPath(userText string) string {
	if name := filenameRE.FindString(userText); name != "" {
		return name
	}
	lower := strings.ToLower(userText)
	stem := slugStem(researchQueryFrom(userText, ""))
	if stem == "" {
		stem = "research_prototype"
	}
	if strings.Contains(lower, "vaccine") || strings.Contains(lower, "therap") {
		if !strings.Contains(stem, "research") {
			stem += "_research"
		}
	}
	switch {
	case strings.Contains(lower, "notebook") || strings.Contains(lower, "ipynb"):
		return stem + ".ipynb"
	case strings.Contains(lower, "chempy") || strings.Contains(lower, "python") || strings.Contains(lower, "script"):
		return stem + ".py"
	case strings.Contains(lower, "golang") || strings.Contains(lower, " go "):
		return stem + ".go"
	default:
		return stem + ".py"
	}
}

func slugStem(text string) string {
	var parts []string
	for _, field := range strings.Fields(strings.ToLower(text)) {
		var b strings.Builder
		for _, r := range field {
			if unicode.IsLetter(r) || unicode.IsDigit(r) {
				b.WriteRune(r)
			}
		}
		tok := b.String()
		if tok == "" || len(tok) < 2 {
			continue
		}
		parts = append(parts, tok)
		if len(parts) >= 4 {
			break
		}
	}
	if len(parts) == 0 {
		return ""
	}
	stem := strings.Join(parts, "_")
	if len(stem) > 48 {
		stem = stem[:48]
	}
	return stem
}

func extractFencedCode(text string) string {
	text = strings.TrimSpace(text)
	if text == "" {
		return ""
	}
	if m := fenceRE.FindStringSubmatch(text); len(m) == 2 {
		return strings.TrimSpace(m[1])
	}
	if looksLikeSource(text) {
		return text
	}
	return ""
}

func looksLikeSource(text string) bool {
	q := strings.TrimSpace(text)
	if q == "" {
		return false
	}
	if strings.HasPrefix(q, "#!/") || strings.HasPrefix(q, "\"\"\"") || strings.HasPrefix(q, "'''") {
		return true
	}
	lines := strings.Split(q, "\n")
	hits := 0
	for i, line := range lines {
		if i > 40 {
			break
		}
		trim := strings.TrimSpace(line)
		switch {
		case strings.HasPrefix(trim, "import ") || strings.HasPrefix(trim, "from "):
			hits += 2
		case strings.HasPrefix(trim, "def ") || strings.HasPrefix(trim, "class "):
			hits += 2
		case strings.HasPrefix(trim, "function ") || strings.HasPrefix(trim, "package "):
			hits += 2
		case strings.HasPrefix(trim, "#") && i < 8:
			hits++
		}
	}
	return hits >= 2
}

func findWriteTool(tools []openaiTool) (openaiTool, bool) {
	prefer := []string{"write", "write_file", "writefile", "create_file", "createfile"}
	byName := map[string]openaiTool{}
	for _, tool := range tools {
		name := strings.ToLower(strings.TrimSpace(tool.Function.Name))
		if name == "" {
			continue
		}
		byName[name] = tool
	}
	for _, name := range prefer {
		if tool, ok := byName[name]; ok {
			return tool, true
		}
	}
	for name, tool := range byName {
		if strings.Contains(name, "todo") {
			continue
		}
		if strings.Contains(name, "write") || strings.Contains(name, "create_file") {
			return tool, true
		}
	}
	return openaiTool{}, false
}

func writeToolArgKeys(tool openaiTool) (pathKey, contentKey string) {
	pathKey, contentKey = "file_path", "content"
	if len(tool.Function.Parameters) == 0 {
		return pathKey, contentKey
	}
	var schema struct {
		Properties map[string]json.RawMessage `json:"properties"`
		Required   []string                   `json:"required"`
	}
	if json.Unmarshal(tool.Function.Parameters, &schema) != nil || len(schema.Properties) == 0 {
		return pathKey, contentKey
	}
	pathAliases := []string{"file_path", "filePath", "path", "filename", "file", "target_file", "targetFile"}
	contentAliases := []string{"content", "contents", "text", "body", "file_text", "fileText"}
	if key := firstSchemaKey(schema.Properties, schema.Required, pathAliases); key != "" {
		pathKey = key
	}
	if key := firstSchemaKey(schema.Properties, schema.Required, contentAliases); key != "" {
		contentKey = key
	}
	return pathKey, contentKey
}

func firstSchemaKey(props map[string]json.RawMessage, required, aliases []string) string {
	lower := map[string]string{}
	for key := range props {
		lower[strings.ToLower(key)] = key
	}
	for _, req := range required {
		if orig, ok := lower[strings.ToLower(req)]; ok {
			for _, alias := range aliases {
				if strings.EqualFold(req, alias) {
					return orig
				}
			}
		}
	}
	for _, alias := range aliases {
		if orig, ok := lower[strings.ToLower(alias)]; ok {
			return orig
		}
	}
	return ""
}

func synthesizeWriteCall(tool openaiTool, path, content string) openaiToolCall {
	pathKey, contentKey := writeToolArgKeys(tool)
	raw, _ := json.Marshal(map[string]string{
		pathKey:    path,
		contentKey: content,
	})
	return openaiToolCall{
		ID:    newOpenAIID("call_"),
		Type:  "function",
		Index: 0,
		Function: openaiToolFunction{
			Name:      tool.Function.Name,
			Arguments: string(raw),
		},
	}
}

func prependHit(hits []RankedChunk, ev RankedChunk) []RankedChunk {
	out := make([]RankedChunk, 0, len(hits)+1)
	out = append(out, ev)
	for _, hit := range hits {
		if hit.Chunk.DOI != "" && hit.Chunk.DOI == ev.Chunk.DOI {
			continue
		}
		out = append(out, hit)
	}
	return out
}

func selectArtifactEvidence(userText, transcript string, hits []RankedChunk) RankedChunk {
	blob := userText + " " + transcript
	var dois []string
	for _, raw := range doiRE.FindAllString(blob, 8) {
		dois = append(dois, strings.TrimRight(raw, ".,;"))
	}
	for _, doi := range dois {
		for _, hit := range hits {
			if hit.Chunk.DOI != "" && (strings.Contains(hit.Chunk.DOI, doi) || strings.Contains(doi, hit.Chunk.DOI)) {
				return hit
			}
		}
	}
	query := strings.ToLower(researchQueryFrom(userText, transcript))
	terms := strings.Fields(query)
	best := RankedChunk{}
	bestScore := -1.0
	for _, hit := range hits {
		blob := strings.ToLower(hit.Chunk.Title + " " + hit.Chunk.Text + " " + hit.Chunk.DOI)
		score := hit.Score
		for _, term := range terms {
			if len(term) >= 4 && strings.Contains(blob, term) {
				score += 0.5
			}
		}
		if strings.EqualFold(hit.Chunk.Source, "scienceopen") && hit.Chunk.DOI != "" {
			score += 0.1
		}
		if score > bestScore {
			bestScore = score
			best = hit
		}
	}
	if bestScore >= 0 {
		return best
	}
	if len(dois) > 0 {
		return RankedChunk{Chunk: KnowledgeChunk{
			DOI:   dois[0],
			Title: "ScienceOpen preprint",
			Text:  singleLine(blob, 700),
			URL:   scienceOpenURL(dois[0]),
		}}
	}
	return RankedChunk{}
}

func artifactAssistantNote(path string, hits []RankedChunk) string {
	if len(hits) > 0 && strings.TrimSpace(hits[0].Chunk.Title+hits[0].Chunk.DOI) != "" {
		hit := hits[0]
		doi := strings.TrimSpace(hit.Chunk.DOI)
		title := singleLine(hit.Chunk.Title, 140)
		if doi == "" {
			doi = "DOI n/a"
		}
		if title == "" {
			title = "ScienceOpen preprint"
		}
		return fmt.Sprintf("Retrieved ScienceOpen preprint %s (%s). Drafting %s from that evidence for OpenCode to write. Computational research only, not a therapy.", title, doi, path)
	}
	return fmt.Sprintf("Drafting %s from retrieved research evidence for OpenCode to write. Computational research only, not a therapy.", path)
}

func (m *StellaModel) BuildCodePrompt(question string, local Prediction, hits []RankedChunk) string {
	var b strings.Builder
	b.WriteString("You are Stella V. The user asked for a file or script. Stella retrieves ScienceOpen preprints and pushes that evidence; you must WRITE THE REQUESTED PROGRAM, not summarize the papers.\n")
	b.WriteString("Rules:\n")
	b.WriteString("1. Output only the complete file in one markdown code fence.\n")
	b.WriteString("2. Ground comments, constants, and model structure in the passages. Cite the DOI in a module docstring.\n")
	b.WriteString("3. This is a computational research prototype. Not a real vaccine, not manufacturing, not diagnosis or treatment.\n")
	b.WriteString("4. If the user asked for Chempy, use the chempy library (Reaction / ReactionSystem, optional ODE kinetics) with placeholder rates.\n")
	b.WriteString("5. Use placeholder sequences and toy kinetics only. No wet-lab protocol.\n")
	b.WriteString("6. If evidence is thin, still write a clearly labeled prototype and say so in comments.\n\n")
	b.WriteString("User request:\n")
	b.WriteString(strings.TrimSpace(question))
	b.WriteByte('\n')
	if strings.TrimSpace(local.Text) != "" && local.Source != "fallback" && local.Source != "empty" {
		b.WriteString("\nLocal tensor-memory draft (optional):\n")
		b.WriteString(singleLine(local.Text, 280))
		b.WriteByte('\n')
	}
	b.WriteString("\nScienceOpen preprint passages:\n")
	b.WriteString(formatRankedPassages(hits, 4))
	b.WriteString("\n\nWrite the file now. Output only code.")
	return b.String()
}

func fallbackResearchScript(userText string, hits []RankedChunk, transcript ...string) string {
	trans := ""
	if len(transcript) > 0 {
		trans = transcript[0]
	}
	ev := selectArtifactEvidence(userText, trans, hits)
	lower := strings.ToLower(userText)
	title, doi, url, passage := "unindexed ScienceOpen preprint", "n/a", "", "No matching preprint was in tensor memory."
	if strings.TrimSpace(ev.Chunk.Title) != "" {
		title = singleLine(ev.Chunk.Title, 180)
	}
	if strings.TrimSpace(ev.Chunk.DOI) != "" {
		doi = strings.TrimSpace(ev.Chunk.DOI)
	}
	url = strings.TrimSpace(ev.Chunk.URL)
	if strings.TrimSpace(ev.Chunk.Text) != "" {
		passage = singleLine(ev.Chunk.Text, 700)
	}
	if strings.Contains(lower, "chempy") || (strings.Contains(lower, "python") && (strings.Contains(lower, "vaccine") || strings.Contains(lower, "kinetic") || strings.Contains(lower, "mrna"))) {
		return chempyResearchStub(title, doi, url, passage)
	}
	return pythonResearchStub(title, doi, url, passage)
}

func pythonResearchStub(title, doi, url, passage string) string {
	return fmt.Sprintf(`#!/usr/bin/env python3
"""Computational research prototype grounded in a ScienceOpen preprint.

Title: %s
DOI: %s
URL: %s

Passage (unreviewed preprint):
    %s

This is NOT a vaccine, NOT a wet-lab protocol, and NOT clinical advice.
Stella V only retrieved evidence; this file is a research sketch for local experimentation.
"""

from __future__ import annotations

EVIDENCE = {
    "title": %q,
    "doi": %q,
    "url": %q,
    "unreviewed": True,
}


def main() -> None:
    print("ScienceOpen DOI:", EVIDENCE["doi"])
    print("Title:", EVIDENCE["title"])
    print("Unreviewed preprint. Computational sketch only.")


if __name__ == "__main__":
    main()
`, title, doi, url, passage, title, doi, url)
}

func chempyResearchStub(title, doi, url, passage string) string {
	return fmt.Sprintf(`#!/usr/bin/env python3
"""Toy Chempy kinetics for a personalized mRNA cancer-vaccine *research sketch*.

Grounded in ScienceOpen preprint:
    %s
    DOI: %s
    URL: %s

Passage (unreviewed):
    %s

This is NOT a vaccine, NOT manufacturing instructions, and NOT medical advice.
Species and rate constants are placeholders for a computational prototype:
    mRNA -> antigen protein -> immune-activation marker
"""

from __future__ import annotations

DOI = %q

try:
    from chempy import Reaction, ReactionSystem
except ImportError as exc:  # pragma: no cover - optional dependency
    raise SystemExit("Install chempy to run this research sketch: pip install chempy") from exc


def build_system() -> ReactionSystem:
    # Placeholder first-order rates (1/hour). Not fitted to data.
    reactions = [
        Reaction({"mRNA": 1}, {"mRNA": 1, "antigen": 1}, param=0.12, name="translation"),
        Reaction({"mRNA": 1}, {}, param=0.05, name="mrna_decay"),
        Reaction({"antigen": 1}, {"signal": 1}, param=0.03, name="presentation"),
        Reaction({"antigen": 1}, {}, param=0.02, name="antigen_clearance"),
        Reaction({"signal": 1}, {}, param=0.01, name="signal_decay"),
    ]
    return ReactionSystem(reactions, "mRNA antigen signal")


def main() -> None:
    rsys = build_system()
    print("ScienceOpen DOI:", DOI)
    print("Unreviewed preprint. Toy ReactionSystem only.")
    print(rsys)
    try:
        import numpy as np
        from chempy.kinetics.ode import get_odesys

        odesys, _extra = get_odesys(rsys)
        tout = np.linspace(0.0, 48.0, 97)
        result, _info = odesys.integrate(tout, {"mRNA": 1.0, "antigen": 0.0, "signal": 0.0})
        print("t_hours\tmRNA\tantigen\tsignal")
        for t, row in zip(result.x, result.yout):
            print(f"{float(t):.2f}\t{row[0]:.6f}\t{row[1]:.6f}\t{row[2]:.6f}")
    except Exception as exc:
        print("ODE integrate skipped:", exc)


if __name__ == "__main__":
    main()
`, title, doi, url, passage, doi)
}

File: Python/artifact.py
#!/usr/bin/env python3
"""Detect script/file requests and turn retrieved ScienceOpen evidence into artifacts.

Stella V does not write files. When OpenCode (or another client) sends a write
tool, we draft the file from preprint passages and return a tool call.
"""

from __future__ import annotations

import json
import re
import uuid
from typing import Any, Sequence

from tensor_engine import RetrievedChunk

DOI_RE = re.compile(r"10\.\d{4,9}/[-._;()/:A-Za-z0-9]+")
FILENAME_RE = re.compile(r"(?i)\b(?:[\w.-]+/)*[\w.-]+\.(py|r|js|ts|go|rs|ipynb|sh|rb|jl|txt)\b")
FENCE_RE = re.compile(r"```(?:[A-Za-z0-9_+-]+)?\s*\n(.*?)```", re.DOTALL)

_ARTIFACT_VERBS = (
    "write", "create", "generate", "make", "draft", "scaffold", "emit", "produce",
)
_ARTIFACT_NOUNS = (
    "script", "python", "chempy", "notebook", "module", "program", "source file", ".py", "code",
)
_STOP = {
    "write", "writing", "create", "creating", "generate", "generating", "make", "making",
    "draft", "scaffold", "python", "chempy", "script", "scripts", "file", "files", "code",
    "program", "now", "please", "based", "research", "the", "a", "an", "to", "for", "begin",
    "beginning", "start", "starting", "do", "that", "this", "with", "using", "use", "want",
    "wants", "user", "preprint", "passage", "passages", "therefore", "answer", "from",
    "found", "in", "of", "and", "or", "on", "it", "is", "be", "can", "should", "able",
    "just", "so", "those", "these", "i", "will", "get", "numpy", "pandas", "scipy",
    "jupyter", "notebook",
}
_MEDICAL_HINTS = (
    "cancer", "vaccine", "mrna", "rna", "immun", "oncolog", "tumor", "tumour",
    "antigen", "peptide", "trial", "therap", "patient", "clinic", "epidem",
)


def looks_like_artifact_request(text: str) -> bool:
    q = (text or "").lower()
    if not q.strip():
        return False
    has_verb = any(v in q for v in _ARTIFACT_VERBS)
    has_noun = any(n in q for n in _ARTIFACT_NOUNS)
    if has_verb and has_noun:
        return True
    if "write the script" in q or "script now" in q:
        return True
    return bool(FILENAME_RE.search(text or "") and has_verb)


def research_query_from(user_text: str, transcript: str = "") -> str:
    cleaned = _strip_boilerplate(user_text)
    if len(cleaned.split()) < 2:
        cleaned = _strip_boilerplate(transcript)
    if len(cleaned.split()) < 2:
        cleaned = _keep_medical(user_text + " " + transcript)
    doi_match = DOI_RE.search((transcript or "") + " " + (user_text or ""))
    if doi_match:
        cleaned = (cleaned + " " + doi_match.group(0).rstrip(".,;")).strip()
    cleaned = " ".join(cleaned.split())
    return cleaned or (user_text or "").strip()


def _strip_boilerplate(text: str) -> str:
    kept: list[str] = []
    for tok in (text or "").lower().split():
        tok = tok.strip(".,;:!?\"'`")
        if not tok or tok in _STOP:
            continue
        if FILENAME_RE.match(tok):
            continue
        kept.append(tok)
    return " ".join(kept)


def _keep_medical(text: str) -> str:
    kept: list[str] = []
    seen: set[str] = set()
    for tok in (text or "").lower().split():
        tok = tok.strip(".,;:!?\"'`")
        if tok in seen:
            continue
        if any(h in tok for h in _MEDICAL_HINTS):
            kept.append(tok)
            seen.add(tok)
    return " ".join(kept)


def infer_artifact_path(user_text: str) -> str:
    named = FILENAME_RE.search(user_text or "")
    if named:
        return named.group(0)
    lower = (user_text or "").lower()
    stem = _slug_stem(research_query_from(user_text, "")) or "research_prototype"
    if ("vaccine" in lower or "therap" in lower) and "research" not in stem:
        stem += "_research"
    if "notebook" in lower or "ipynb" in lower:
        return stem + ".ipynb"
    if "chempy" in lower or "python" in lower or "script" in lower:
        return stem + ".py"
    return stem + ".py"


def _slug_stem(text: str) -> str:
    parts: list[str] = []
    for field in (text or "").lower().split():
        tok = "".join(ch for ch in field if ch.isalnum())
        if len(tok) < 2:
            continue
        parts.append(tok)
        if len(parts) >= 4:
            break
    stem = "_".join(parts)
    return stem[:48]


def extract_fenced_code(text: str) -> str:
    text = (text or "").strip()
    if not text:
        return ""
    match = FENCE_RE.search(text)
    if match:
        return match.group(1).strip()
    if looks_like_source(text):
        return text
    return ""


def looks_like_source(text: str) -> bool:
    q = (text or "").strip()
    if not q:
        return False
    if q.startswith("#!/") or q.startswith('"""') or q.startswith("'''"):
        return True
    hits = 0
    for i, line in enumerate(q.splitlines()):
        if i > 40:
            break
        trim = line.strip()
        if trim.startswith("import ") or trim.startswith("from "):
            hits += 2
        elif trim.startswith("def ") or trim.startswith("class "):
            hits += 2
        elif trim.startswith("function ") or trim.startswith("package "):
            hits += 2
        elif trim.startswith("#") and i < 8:
            hits += 1
    return hits >= 2


def find_write_tool(tools: Sequence[dict[str, Any]] | None) -> dict[str, Any] | None:
    prefer = ("write", "write_file", "writefile", "create_file", "createfile")
    by_name: dict[str, dict[str, Any]] = {}
    for tool in tools or []:
        fn = tool.get("function") if isinstance(tool, dict) else None
        if not isinstance(fn, dict):
            continue
        name = str(fn.get("name") or "").strip().lower()
        if name:
            by_name[name] = tool
    for name in prefer:
        if name in by_name:
            return by_name[name]
    for name, tool in by_name.items():
        if "todo" in name:
            continue
        if "write" in name or "create_file" in name:
            return tool
    return None


def write_tool_arg_keys(tool: dict[str, Any]) -> tuple[str, str]:
    path_key, content_key = "file_path", "content"
    fn = tool.get("function") if isinstance(tool, dict) else None
    params = (fn or {}).get("parameters") or {}
    if isinstance(params, str):
        try:
            params = json.loads(params)
        except json.JSONDecodeError:
            params = {}
    props = params.get("properties") if isinstance(params, dict) else None
    required = params.get("required") if isinstance(params, dict) else None
    if not isinstance(props, dict):
        return path_key, content_key
    path_key = _first_schema_key(props, required, (
        "file_path", "filePath", "path", "filename", "file", "target_file", "targetFile",
    )) or path_key
    content_key = _first_schema_key(props, required, (
        "content", "contents", "text", "body", "file_text", "fileText",
    )) or content_key
    return path_key, content_key


def _first_schema_key(props: dict[str, Any], required: Any, aliases: tuple[str, ...]) -> str:
    lower = {str(k).lower(): str(k) for k in props}
    if isinstance(required, list):
        for req in required:
            orig = lower.get(str(req).lower())
            if orig and any(str(req).lower() == alias.lower() for alias in aliases):
                return orig
    for alias in aliases:
        if alias.lower() in lower:
            return lower[alias.lower()]
    return ""


def synthesize_write_call(tool: dict[str, Any], path: str, content: str) -> dict[str, Any]:
    fn = tool.get("function") if isinstance(tool, dict) else {}
    name = str((fn or {}).get("name") or "write")
    path_key, content_key = write_tool_arg_keys(tool)
    return {
        "id": "call_" + uuid.uuid4().hex[:12],
        "type": "function",
        "index": 0,
        "function": {
            "name": name,
            "arguments": json.dumps({path_key: path, content_key: content}),
        },
    }


def select_artifact_evidence(
    user_text: str,
    hits: Sequence[RetrievedChunk],
    transcript: str = "",
) -> RetrievedChunk | None:
    blob = f"{user_text} {transcript}"
    dois = [m.rstrip(".,;") for m in DOI_RE.findall(blob)]
    for doi in dois:
        for hit in hits:
            if hit.doi and (doi in hit.doi or hit.doi in doi):
                return hit
    query = research_query_from(user_text, transcript).lower()
    terms = query.split()
    best: RetrievedChunk | None = None
    best_score = -1.0
    for hit in hits:
        text_blob = f"{hit.title} {hit.text} {hit.doi}".lower()
        score = float(hit.score)
        for term in terms:
            if len(term) >= 4 and term in text_blob:
                score += 0.5
        if (hit.source or "").lower() == "scienceopen" and hit.doi:
            score += 0.1
        if score > best_score:
            best_score = score
            best = hit
    if best is not None:
        return best
    if dois:
        return RetrievedChunk(
            text=" ".join(blob.split())[:700],
            score=0.0,
            title="ScienceOpen preprint",
            doi=dois[0],
            url=f"https://www.scienceopen.com/hosted-document?doi={dois[0]}",
            source="scienceopen",
        )
    return None


def artifact_assistant_note(path: str, hits: Sequence[RetrievedChunk]) -> str:
    hit = hits[0] if hits else None
    if hit and (hit.title or hit.doi):
        doi = hit.doi or "DOI n/a"
        title = " ".join((hit.title or "").split())[:140] or "ScienceOpen preprint"
        return (
            f"Retrieved ScienceOpen preprint {title} ({doi}). "
            f"Drafting {path} from that evidence for OpenCode to write. "
            "Computational research only, not a therapy."
        )
    return (
        f"Drafting {path} from retrieved research evidence for OpenCode to write. "
        "Computational research only, not a therapy."
    )


def build_code_prompt(question: str, hits: Sequence[RetrievedChunk], *, local_draft: str = "") -> str:
    from reasoner import format_passages

    draft = ""
    if (local_draft or "").strip():
        draft = f"\nLocal tensor-memory draft (optional):\n{local_draft.strip()[:280]}\n"
    return (
        "You are Stella V. The user asked for a file or script. Stella retrieves "
        "ScienceOpen preprints and pushes that evidence; you must WRITE THE REQUESTED "
        "PROGRAM, not summarize the papers.\n"
        "Rules:\n"
        "1. Output only the complete file in one markdown code fence.\n"
        "2. Ground comments, constants, and model structure in the passages. Cite the DOI in a module docstring.\n"
        "3. This is a computational research prototype. Not a real vaccine, not manufacturing, not diagnosis or treatment.\n"
        "4. If the user asked for Chempy, use the chempy library (Reaction / ReactionSystem, optional ODE kinetics) with placeholder rates.\n"
        "5. Use placeholder sequences and toy kinetics only. No wet-lab protocol.\n"
        "6. If evidence is thin, still write a clearly labeled prototype and say so in comments.\n\n"
        f"User request:\n{question.strip()}\n"
        f"{draft}\n"
        f"ScienceOpen preprint passages:\n{format_passages(hits, limit=4)}\n\n"
        "Write the file now. Output only code."
    )


def fallback_research_script(user_text: str, hits: Sequence[RetrievedChunk], transcript: str = "") -> str:
    title, doi, url, passage = "unindexed ScienceOpen preprint", "n/a", "", "No matching preprint was in tensor memory."
    ev = select_artifact_evidence(user_text, hits, transcript)
    if ev is not None:
        title = " ".join((ev.title or "").split())[:180] or title
        doi = ev.doi or doi
        url = ev.url or ""
        passage = " ".join((ev.text or "").split())[:700] or passage
    lower = (user_text or "").lower()
    if "chempy" in lower or ("python" in lower and any(k in lower for k in ("vaccine", "kinetic", "mrna"))):
        return _chempy_stub(title, doi, url, passage)
    return _python_stub(title, doi, url, passage)


def _python_stub(title: str, doi: str, url: str, passage: str) -> str:
    return f'''#!/usr/bin/env python3
"""Computational research prototype grounded in a ScienceOpen preprint.

Title: {title}
DOI: {doi}
URL: {url}

Passage (unreviewed preprint):
    {passage}

This is NOT a vaccine, NOT a wet-lab protocol, and NOT clinical advice.
Stella V only retrieved evidence; this file is a research sketch for local experimentation.
"""

from __future__ import annotations

EVIDENCE = {{
    "title": {title!r},
    "doi": {doi!r},
    "url": {url!r},
    "unreviewed": True,
}}


def main() -> None:
    print("ScienceOpen DOI:", EVIDENCE["doi"])
    print("Title:", EVIDENCE["title"])
    print("Unreviewed preprint. Computational sketch only.")


if __name__ == "__main__":
    main()
'''


def _chempy_stub(title: str, doi: str, url: str, passage: str) -> str:
    return f'''#!/usr/bin/env python3
"""Toy Chempy kinetics for a personalized mRNA cancer-vaccine *research sketch*.

Grounded in ScienceOpen preprint:
    {title}
    DOI: {doi}
    URL: {url}

Passage (unreviewed):
    {passage}

This is NOT a vaccine, NOT manufacturing instructions, and NOT medical advice.
Species and rate constants are placeholders for a computational prototype:
    mRNA -> antigen protein -> immune-activation marker
"""

from __future__ import annotations

DOI = {doi!r}

try:
    from chempy import Reaction, ReactionSystem
except ImportError as exc:  # pragma: no cover - optional dependency
    raise SystemExit("Install chempy to run this research sketch: pip install chempy") from exc


def build_system() -> ReactionSystem:
    # Placeholder first-order rates (1/hour). Not fitted to data.
    reactions = [
        Reaction({{"mRNA": 1}}, {{"mRNA": 1, "antigen": 1}}, param=0.12, name="translation"),
        Reaction({{"mRNA": 1}}, {{}}, param=0.05, name="mrna_decay"),
        Reaction({{"antigen": 1}}, {{"signal": 1}}, param=0.03, name="presentation"),
        Reaction({{"antigen": 1}}, {{}}, param=0.02, name="antigen_clearance"),
        Reaction({{"signal": 1}}, {{}}, param=0.01, name="signal_decay"),
    ]
    return ReactionSystem(reactions, "mRNA antigen signal")


def main() -> None:
    rsys = build_system()
    print("ScienceOpen DOI:", DOI)
    print("Unreviewed preprint. Toy ReactionSystem only.")
    print(rsys)
    try:
        import numpy as np
        from chempy.kinetics.ode import get_odesys

        odesys, _extra = get_odesys(rsys)
        tout = np.linspace(0.0, 48.0, 97)
        result, _info = odesys.integrate(tout, {{"mRNA": 1.0, "antigen": 0.0, "signal": 0.0}})
        print("t_hours\\tmRNA\\tantigen\\tsignal")
        for t, row in zip(result.x, result.yout):
            print(f"{{float(t):.2f}}\\t{{row[0]:.6f}}\\t{{row[1]:.6f}}\\t{{row[2]:.6f}}")
    except Exception as exc:
        print("ODE integrate skipped:", exc)


if __name__ == "__main__":
    main()
'''
